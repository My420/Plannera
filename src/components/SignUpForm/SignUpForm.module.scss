@import '../../sass/variables.scss';
@import '../../sass/mixins.scss';

.form {
  @include auth-form;
}

.title {
  margin-bottom: 10px;
  font-weight: 700;
  text-align: center;
  color: $fonts-auth-color;
}

.signup,
.controls {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  width: 300px;
  margin: 0 auto;

  border: none;
}

.controls {
  display: block;
}

.signup {
  margin-bottom: 30px;
}

.link {
  @include text-link;
  display: block;
  width: 70px;
  margin: 0 auto;
  margin-top: 10px;

  text-align: center;
  color: $fonts-auth-color;
}

@media (max-width: $mobile-large) {
  .signup,
  .controls {
    width: 280px;
  }

  .title {
    margin-bottom: 0px;
  }
}
