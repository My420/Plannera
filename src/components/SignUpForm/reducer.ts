import { IAuthField, ISignUpState, IAuthAction } from '../../types/signUpForm';
import {
  EMAIL, PASSWORD, CONFIRM, FIRST_NAME, LAST_NAME,
} from '../../utils/constant';
import validateEmail from '../../utils/validateEmail';
import validatePassword from '../../utils/validatePassword';
import validateConfirmField from '../../utils/validateConfirmField';
import validateNameField from '../../utils/validateNameField';

export const field: IAuthField = { value: '', isValid: false, error: '' };
export const initialState: ISignUpState = {
  [EMAIL]: { ...field },
  [PASSWORD]: { ...field },
  [CONFIRM]: { ...field },
  [FIRST_NAME]: { ...field },
  [LAST_NAME]: { ...field },
};

const reducer = (state: ISignUpState, action: IAuthAction) => {
  const {
    type,
    payload: { value },
  } = action;

  const newState = (isValid: boolean, error: string): ISignUpState => ({
    ...state,
    [type]: { value, isValid, error },
  });

  switch (type) {
    case EMAIL: {
      const { isValid, error } = validateEmail(value);
      return newState(isValid, error);
    }
    case PASSWORD: {
      const passwordStatus = validatePassword(value);
      const confirmStatus = validateConfirmField(state[CONFIRM].value, value);
      if (!passwordStatus.isValid || (passwordStatus.isValid && confirmStatus.isValid)) {
        return newState(passwordStatus.isValid, passwordStatus.error);
      }
      return {
        ...newState(passwordStatus.isValid, passwordStatus.error),
        [CONFIRM]: { ...state[CONFIRM], ...confirmStatus },
      };
    }
    case CONFIRM: {
      const { isValid, error } = validateConfirmField(value, state[PASSWORD].value);
      return newState(isValid, error);
    }
    case FIRST_NAME: {
      const { isValid, error } = validateNameField(value);
      return newState(isValid, error);
    }
    case LAST_NAME: {
      const { isValid, error } = validateNameField(value);
      return newState(isValid, error);
    }
    default:
      return state;
  }
};

export default reducer;
