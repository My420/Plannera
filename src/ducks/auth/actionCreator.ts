import { ISignUpFormData, ISignInFormData } from '../../types/signUpForm';
import {
  ISignUpRequestAction,
  ISignUpErrorAction,
  ISignInRequestAction,
  ISignInSuccessAction,
  ISignInErrorAction,
  ISignOutRequestAction,
  ISignOutSuccessAction,
  ISignOutErrorAction,
  IAuthData,
} from './types';
import {
  SIGN_UP_REQUEST,
  SIGN_UP_ERROR,
  SIGN_IN_REQUEST,
  SIGN_IN_SUCCESS,
  SIGN_IN_ERROR,
  SIGN_OUT_REQUEST,
  SIGN_OUT_SUCCESS,
  SIGN_OUT_ERROR,
} from './constant';

export const signUp = (data: ISignUpFormData): ISignUpRequestAction => ({
  type: SIGN_UP_REQUEST,
  payload: data,
});

export const signUpError = (error: string): ISignUpErrorAction => ({
  type: SIGN_UP_ERROR,
  payload: error,
});

export const signIn = (data: ISignInFormData): ISignInRequestAction => ({
  type: SIGN_IN_REQUEST,
  payload: data,
});

export const signInSuccess = (data: IAuthData): ISignInSuccessAction => ({
  type: SIGN_IN_SUCCESS,
  payload: data,
});

export const signInError = (error: string): ISignInErrorAction => ({
  type: SIGN_IN_ERROR,
  payload: error,
});

export const signOut = (): ISignOutRequestAction => ({
  type: SIGN_OUT_REQUEST,
  payload: null,
});

export const signOutSuccess = (): ISignOutSuccessAction => ({
  type: SIGN_OUT_SUCCESS,
  payload: null,
});

export const signOutError = (error: string): ISignOutErrorAction => ({
  type: SIGN_OUT_ERROR,
  payload: error,
});
