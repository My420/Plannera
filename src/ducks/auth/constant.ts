import { fb as firebase } from '../../firebase/config';

export const moduleName = 'auth';
export const SIGN_UP_REQUEST = 'Plannera/auth/SIGN_UP_REQUEST';
export const SIGN_UP_SUCCESS = 'Plannera/auth/SIGN_UP_SUCCESS';
export const SIGN_UP_ERROR = 'Plannera/auth/SIGN_UP_ERROR';
export const SIGN_IN_REQUEST = 'Plannera/auth/SIGN_IN_REQUEST';
export const SIGN_IN_SUCCESS = 'Plannera/auth/SIGN_IN_SUCCESS';
export const SIGN_IN_ERROR = 'Plannera/auth/SIGN_IN_ERROR';
export const SIGN_OUT_REQUEST = 'Plannera/auth/SIGN_OUT_REQUEST';
export const SIGN_OUT_SUCCESS = 'Plannera/auth/SIGN_OUT_SUCCESS';
export const SIGN_OUT_ERROR = 'Plannera/auth/SIGN_OUT_ERROR';
export const AUTH_CLEAR_ERROR = 'Plannera/auth/AUTH_CLEAR_ERROR';

export const auth = firebase.auth();
