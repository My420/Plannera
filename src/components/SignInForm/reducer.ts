import { ISignInState, IAuthAction, IAuthField } from '../../types/signUpForm';
import { EMAIL, PASSWORD } from '../../utils/constant';
import validateEmail from '../../utils/validateEmail';
import validatePassword from '../../utils/validatePassword';

export const field: IAuthField = { value: '', isValid: false, error: '' };
export const initialState: ISignInState = {
  [EMAIL]: { ...field },
  [PASSWORD]: { ...field },
};

const reducer = (state: ISignInState, action: IAuthAction) => {
  const {
    type,
    payload: { value },
  } = action;

  const newState = (isValid: boolean, error: string): ISignInState => ({
    ...state,
    [type]: { value, isValid, error },
  });

  switch (type) {
    case EMAIL: {
      const { isValid, error } = validateEmail(value);
      return newState(isValid, error);
    }
    case PASSWORD: {
      const { isValid, error } = validatePassword(value);
      return newState(isValid, error);
    }
    default:
      return state;
  }
};

export default reducer;
