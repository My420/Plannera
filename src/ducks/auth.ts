import { Record } from 'immutable';
import { put, takeEvery, all } from 'redux-saga/effects';
import { ISignUpFormData, ISignInFormData } from '../types/signUpForm';
import { IUser } from '../types/user';
import { EMAIL, LAST_NAME, FIRST_NAME } from '../utils/constant';

export const moduleName = 'auth';
export const SIGN_UP_REQUEST = 'Plannera/auth/SIGN_UP_REQUEST';
export const SIGN_UP_SUCCESS = 'Plannera/auth/SIGN_UP_SUCCESS';
export const SIGN_UP_ERROR = 'Plannera/auth/SIGN_UP_ERROR';
export const SIGN_IN_REQUEST = 'Plannera/auth/SIGN_IN_REQUEST';

/**
 * types
 */

export interface ISignUpRequestAction {
  type: typeof SIGN_UP_REQUEST;
  payload: ISignUpFormData;
}
export interface ISignUpSuccessAction {
  type: typeof SIGN_UP_SUCCESS;
  payload: IUser;
}
export interface ISignUpErrorAction {
  type: typeof SIGN_UP_ERROR;
  payload: string;
}
export interface ISignInRequestAction {
  type: typeof SIGN_IN_REQUEST;
  payload: ISignInFormData;
}

export type AuthActionTypes =
  | ISignUpRequestAction
  | ISignUpSuccessAction
  | ISignUpErrorAction
  | ISignInRequestAction;

export interface IReducerInitialState extends IUser {
  error: null | string;
  loading: boolean;
}

/**
 * Reducer
 */

const initialState: IReducerInitialState = {
  userID: null,
  email: null,
  firstName: null,
  lastName: null,
  initial: null,
  error: null,
  loading: false,
};

const ReducerRecord = Record(initialState);
const reducerState = new ReducerRecord();

const reducer = (state = reducerState, action: AuthActionTypes): typeof reducerState => {
  switch (action.type) {
    case SIGN_UP_REQUEST:
      return state.set('loading', true);
    case SIGN_UP_SUCCESS: {
      const {
        userID, email, lastName, firstName, initial,
      } = action.payload;
      return state
        .set('userID', userID)
        .set('email', email)
        .set('firstName', firstName)
        .set('lastName', lastName)
        .set('initial', initial)
        .set('loading', false)
        .set('error', null);
    }
    case SIGN_UP_ERROR: {
      const error = action.payload;
      return state
        .set('userID', null)
        .set('email', null)
        .set('firstName', null)
        .set('lastName', null)
        .set('initial', null)
        .set('loading', false)
        .set('error', error);
    }
    default:
      return state;
  }
};

export default reducer;

/**
 * Action Creator
 */

export const signUp = (data: ISignUpFormData): ISignUpRequestAction => ({
  type: SIGN_UP_REQUEST,
  payload: data,
});

export const signUpSuccess = (data: IUser): ISignUpSuccessAction => ({
  type: SIGN_UP_SUCCESS,
  payload: data,
});

export const signIn = (data: ISignInFormData): ISignInRequestAction => ({
  type: SIGN_IN_REQUEST,
  payload: data,
});

/**
 *  Saga
 */

export function* registerUser(action: ISignUpRequestAction) {
  const { payload } = action;
  const { [EMAIL]: email, [FIRST_NAME]: firstName, [LAST_NAME]: lastName } = payload;

  // register new user in firebase get ID

  // create initial
  const initial = (firstName[0] + lastName[0]).toUpperCase();

  // add user profile card

  // dispatch success or error action

  const data: IUser = {
    userID: 'dsdfadsasdad2dsdsa',
    email,
    firstName,
    lastName,
    initial,
  };

  yield put(signUpSuccess(data));
}

export function* loginUser(action: ISignInRequestAction) {
  const { payload } = action;
  const { [EMAIL]: email } = payload;

  // login user in firebase get ID

  // fetch user data from firebase

  // dispatch success or error action

  const data: IUser = {
    userID: 'dsdfadsasdad2dsdsa',
    email,
    firstName: 'firstName',
    lastName: 'lastName',
    initial: 'FL',
  };

  yield put(signUpSuccess(data));
}

export function* authSaga() {
  yield all([takeEvery(SIGN_UP_REQUEST, registerUser), takeEvery(SIGN_IN_REQUEST, loginUser)]);
}
