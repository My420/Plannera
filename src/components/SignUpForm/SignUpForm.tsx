import React, { useReducer } from 'react';
import { useDispatch } from 'react-redux';
import { Dispatch } from 'redux';
import { Link } from 'react-router-dom';
import { IAuthAction } from '../../types/signUpForm';
import {
  EMAIL,
  PASSWORD,
  CONFIRM,
  FIRST_NAME,
  LAST_NAME,
  SIGN_IN_PAGE,
} from '../../utils/constant';
import reducer, { initialState } from './reducer';
import { signUp, clearAuthError } from '../../ducks/auth/actionCreator';
import AppActions from '../../ducks/appActionsType';
import EmailAuthInput from '../Inputs/EmailAuthInput';
import PasswordAuthInput from '../Inputs/PasswordAuthInput';
import ConfirmAuthInput from '../Inputs/ConfirmAuthInput';
import FirstNameAuthInput from '../Inputs/FirstNameAuthInput';
import LastNameAuthInput from '../Inputs/LastNameAuthInput';
import MenuSubmitButton from '../Buttons/MenuSubmitButton';
import styles from './SignUpForm.module.scss';
import AuthRequestBar from '../RequestStatusBar/AuthRequestBar';

export interface ISignUpFormProps {}

const SignUpForm: React.FC<ISignUpFormProps> = () => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const reduxDispatch = useDispatch<Dispatch<AppActions>>();

  const setNewValue = (name: string, value: string): void => {
    const action: IAuthAction = { type: name, payload: { value } };
    dispatch(action);
  };

  const onSubmit = (evt: React.MouseEvent<HTMLButtonElement>): void => {
    evt.preventDefault();
    reduxDispatch(
      signUp({
        [EMAIL]: state[EMAIL].value,
        [PASSWORD]: state[PASSWORD].value,
        [FIRST_NAME]: state[FIRST_NAME].value,
        [LAST_NAME]: state[LAST_NAME].value,
      }),
    );
  };

  const clearError = (): void => {
    reduxDispatch(clearAuthError());
  };

  const isSubmitDisabled: boolean = !(
    state[EMAIL].isValid
    && state[PASSWORD].isValid
    && state[CONFIRM].isValid
    && state[LAST_NAME].isValid
    && state[FIRST_NAME].isValid
  );

  return (
    <form className={styles.form} name="signUpForm" action="signup.php">
      <p className={styles.title}>Create account</p>
      <div className={styles.status}>
        <AuthRequestBar />
      </div>
      <fieldset className={styles.signup}>
        <EmailAuthInput
          extensionContainerClass={styles.input}
          value={state[EMAIL].value}
          isValid={state[EMAIL].isValid}
          error={state[EMAIL].error}
          onInputChange={setNewValue}
        />
        <PasswordAuthInput
          extensionContainerClass={styles.input}
          value={state[PASSWORD].value}
          isValid={state[PASSWORD].isValid}
          error={state[PASSWORD].error}
          onInputChange={setNewValue}
        />
        <ConfirmAuthInput
          extensionContainerClass={styles.input}
          value={state[CONFIRM].value}
          isValid={state[CONFIRM].isValid}
          error={state[CONFIRM].error}
          onInputChange={setNewValue}
        />

        <FirstNameAuthInput
          extensionContainerClass={styles.input}
          value={state[FIRST_NAME].value}
          isValid={state[FIRST_NAME].isValid}
          error={state[FIRST_NAME].error}
          onInputChange={setNewValue}
        />

        <LastNameAuthInput
          extensionContainerClass={styles.input}
          value={state[LAST_NAME].value}
          isValid={state[LAST_NAME].isValid}
          error={state[LAST_NAME].error}
          onInputChange={setNewValue}
        />
      </fieldset>
      <fieldset className={styles.controls}>
        <MenuSubmitButton
          text="Continue"
          name="signUpSubmit"
          onButtonClick={onSubmit}
          disabled={isSubmitDisabled}
        />
        <Link className={styles.link} to={SIGN_IN_PAGE} onClick={clearError}>
          Sign in
        </Link>
      </fieldset>
    </form>
  );
};

export default SignUpForm;
