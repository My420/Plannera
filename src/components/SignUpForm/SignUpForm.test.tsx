import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { render, cleanup, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { moduleName } from '../../ducks/auth/constant';
import {
  EMAIL, PASSWORD, LAST_NAME, FIRST_NAME,
} from '../../utils/constant';
import SignUpForm from './SignUpForm';
import { reducerState } from '../../ducks/auth/reducer';
import validateEmail from '../../utils/validateEmail';
import validatePassword from '../../utils/validatePassword';
import { signUp } from '../../ducks/auth/actionCreator';
import validateConfirmField from '../../utils/validateConfirmField';
import validateNameField from '../../utils/validateNameField';

const mockStore = configureStore();

const renderWithAuthRedux = (component: React.ReactNode, initialState = reducerState) => {
  const store = mockStore({ [moduleName]: initialState });
  return {
    ...render(
      <Provider store={store}>
        <MemoryRouter>{component}</MemoryRouter>
      </Provider>,
    ),
    store,
  };
};

describe('test SignUpForm component', () => {
  afterEach(cleanup);

  const text = {
    email: {
      valid: 'test@test.ru',
      invalid: 'test@test',
      errorMsg: validateEmail('test@test').error,
      emptyMsg: validateEmail('').error,
    },
    password: {
      valid: '12345678',
      invalid: '1234',
      errorMsg: validatePassword('1234').error,
      emptyMsg: validatePassword('').error,
    },
    confirm: {
      valid: '12345678',
      invalid: '12345',
      errorMsg: validateConfirmField('12345678', '12345').error,
      emptyMsg: validateConfirmField('12345', '12345678').error,
    },
    name: {
      valid: 'TestName',
      invalid: 'T',
      errorMsg: validateNameField('T').error,
      emptyMsg: validateNameField('').error,
    },
  };

  describe('component view test', () => {
    test('should match snapshot', () => {
      const { asFragment } = renderWithAuthRedux(<SignUpForm />);
      expect(asFragment()).toMatchSnapshot();
    });

    test('should render element with form name', () => {
      renderWithAuthRedux(<SignUpForm />);
      const element = screen.getByText('Create account');
      expect(element).toBeTruthy();
    });

    test('request status bar should be empty', () => {
      renderWithAuthRedux(<SignUpForm />);
      const element = screen.getByTestId('container');
      expect(element).toBeEmpty();
    });

    test('there should be no error messages', () => {
      renderWithAuthRedux(<SignUpForm />);
      const element = screen.getAllByTestId('error');
      expect(element).toHaveLength(5);
      element.forEach((elem) => {
        expect(elem).toBeEmpty();
      });
    });

    test('should render necessary input fields', () => {
      renderWithAuthRedux(<SignUpForm />);
      const inputs = ['Email', 'Password', 'Confirm password', 'First name', 'Last name'];
      inputs.forEach((p) => {
        const elem = screen.getByPlaceholderText(p);
        expect(elem).toBeTruthy();
      });
    });

    test('submit button should be disabled', () => {
      renderWithAuthRedux(<SignUpForm />);
      const element = screen.getByText('Continue');
      expect(element).toBeTruthy();
      expect(element).toBeDisabled();
    });

    test('should render link on Sign In page', () => {
      renderWithAuthRedux(<SignUpForm />);
      const element = screen.getByText('Sign in');
      expect(element).toHaveAttribute('href', '/signin');
    });
  });

  describe('component logic test', () => {
    test('email input should work correct with valid text ', async () => {
      renderWithAuthRedux(<SignUpForm />);
      const emailInput = screen.getByPlaceholderText('Email');
      await userEvent.type(emailInput, text.email.valid, { allAtOnce: true });

      expect(emailInput).toHaveValue(text.email.valid);
      const emailErrorField = emailInput.previousSibling;
      expect(emailErrorField).toBeEmpty();
    });

    test('email input should work correct with invalid text ', async () => {
      renderWithAuthRedux(<SignUpForm />);
      const emailInput = screen.getByPlaceholderText('Email');

      await userEvent.type(emailInput, text.email.invalid, { allAtOnce: true });
      expect(emailInput).toHaveValue(text.email.invalid);
      const emailErrorField = emailInput.previousSibling;
      expect(emailErrorField).toHaveTextContent(text.email.errorMsg);

      userEvent.clear(emailInput);
      expect(emailInput).toHaveValue('');
      expect(emailErrorField).toHaveTextContent(text.email.emptyMsg);
    });

    test('email input should work correct when invalid text change for valid', async () => {
      renderWithAuthRedux(<SignUpForm />);
      const emailInput = screen.getByPlaceholderText('Email');

      await userEvent.type(emailInput, text.email.invalid, { allAtOnce: true });
      expect(emailInput).toHaveValue(text.email.invalid);
      const emailErrorField = emailInput.previousSibling;
      expect(emailErrorField).toHaveTextContent(text.email.errorMsg);

      userEvent.clear(emailInput);
      await userEvent.type(emailInput, text.email.valid, { allAtOnce: true });
      expect(emailInput).toHaveValue(text.email.valid);
      expect(emailErrorField).toBeEmpty();
    });

    test('password input should work correct with valid text ', async () => {
      renderWithAuthRedux(<SignUpForm />);
      const passwordInput = screen.getByPlaceholderText('Password');
      await userEvent.type(passwordInput, text.password.valid, { allAtOnce: true });

      expect(passwordInput).toHaveValue(text.password.valid);
      const passwordErrorField = passwordInput.previousSibling;
      expect(passwordErrorField).toBeEmpty();
    });

    test('password input should work correct with invalid text ', async () => {
      renderWithAuthRedux(<SignUpForm />);
      const passwordInput = screen.getByPlaceholderText('Password');

      await userEvent.type(passwordInput, text.password.invalid, { allAtOnce: true });
      expect(passwordInput).toHaveValue(text.password.invalid);
      const passwordErrorField = passwordInput.previousSibling;
      expect(passwordErrorField).toHaveTextContent(text.password.errorMsg);

      userEvent.clear(passwordInput);
      expect(passwordInput).toHaveValue('');
      expect(passwordErrorField).toHaveTextContent(text.password.emptyMsg);
    });

    test('password input should work correct when invalid text change for valid', async () => {
      renderWithAuthRedux(<SignUpForm />);
      const passwordInput = screen.getByPlaceholderText('Password');

      await userEvent.type(passwordInput, text.password.invalid, { allAtOnce: true });
      expect(passwordInput).toHaveValue(text.password.invalid);
      const passwordErrorField = passwordInput.previousSibling;
      expect(passwordErrorField).toHaveTextContent(text.password.errorMsg);

      userEvent.clear(passwordInput);
      await userEvent.type(passwordInput, text.password.valid, { allAtOnce: true });
      expect(passwordInput).toHaveValue(text.password.valid);
      expect(passwordErrorField).toBeEmpty();
    });

    test('confirm input should work correct with valid text ', async () => {
      renderWithAuthRedux(<SignUpForm />);
      const passwordInput = screen.getByPlaceholderText('Password');
      await userEvent.type(passwordInput, text.password.valid, { allAtOnce: true });

      const confirmInput = screen.getByPlaceholderText('Confirm password');
      await userEvent.type(confirmInput, text.confirm.valid, { allAtOnce: true });

      expect(confirmInput).toHaveValue(text.confirm.valid);
      const confirmErrorField = confirmInput.previousSibling;
      expect(confirmErrorField).toBeEmpty();
    });

    test('confirm input should work correct with invalid text ', async () => {
      renderWithAuthRedux(<SignUpForm />);
      const passwordInput = screen.getByPlaceholderText('Password');
      await userEvent.type(passwordInput, text.password.valid, { allAtOnce: true });

      const confirmInput = screen.getByPlaceholderText('Confirm password');
      await userEvent.type(confirmInput, text.confirm.invalid, { allAtOnce: true });

      expect(confirmInput).toHaveValue(text.confirm.invalid);
      const confirmErrorField = confirmInput.previousSibling;
      expect(confirmErrorField).toHaveTextContent(text.confirm.errorMsg);

      userEvent.clear(confirmInput);
      expect(confirmInput).toHaveValue('');
      expect(confirmErrorField).toHaveTextContent(text.confirm.emptyMsg);
    });

    test('confirm input should work correct when invalid text change for valid', async () => {
      renderWithAuthRedux(<SignUpForm />);
      const passwordInput = screen.getByPlaceholderText('Password');
      await userEvent.type(passwordInput, text.password.valid, { allAtOnce: true });

      const confirmInput = screen.getByPlaceholderText('Confirm password');
      await userEvent.type(confirmInput, text.confirm.invalid, { allAtOnce: true });

      expect(confirmInput).toHaveValue(text.confirm.invalid);
      const confirmErrorField = confirmInput.previousSibling;
      expect(confirmErrorField).toHaveTextContent(text.confirm.errorMsg);

      userEvent.clear(confirmInput);
      await userEvent.type(confirmInput, text.confirm.valid, { allAtOnce: true });

      expect(confirmInput).toHaveValue(text.confirm.valid);
      expect(confirmErrorField).toBeEmpty();
    });

    test('first name input should work correct with valid text ', async () => {
      renderWithAuthRedux(<SignUpForm />);
      const nameInput = screen.getByPlaceholderText('First name');
      await userEvent.type(nameInput, text.name.valid, { allAtOnce: true });

      expect(nameInput).toHaveValue(text.name.valid);
      const nameErrorField = nameInput.previousSibling;
      expect(nameErrorField).toBeEmpty();
    });

    test('first name input should work correct with invalid text ', async () => {
      renderWithAuthRedux(<SignUpForm />);
      const nameInput = screen.getByPlaceholderText('First name');

      await userEvent.type(nameInput, text.name.invalid, { allAtOnce: true });
      expect(nameInput).toHaveValue(text.name.invalid);
      const nameErrorField = nameInput.previousSibling;
      expect(nameErrorField).toHaveTextContent(text.name.errorMsg);

      userEvent.clear(nameInput);
      expect(nameInput).toHaveValue('');
      expect(nameErrorField).toHaveTextContent(text.name.emptyMsg);
    });

    test('first name input should work correct when invalid text change for valid', async () => {
      renderWithAuthRedux(<SignUpForm />);
      const nameInput = screen.getByPlaceholderText('First name');

      await userEvent.type(nameInput, text.name.invalid, { allAtOnce: true });
      expect(nameInput).toHaveValue(text.name.invalid);
      const nameErrorField = nameInput.previousSibling;
      expect(nameErrorField).toHaveTextContent(text.name.errorMsg);

      userEvent.clear(nameInput);
      await userEvent.type(nameInput, text.name.valid, { allAtOnce: true });
      expect(nameInput).toHaveValue(text.name.valid);
      expect(nameErrorField).toBeEmpty();
    });

    test('last name input should work correct with valid text ', async () => {
      renderWithAuthRedux(<SignUpForm />);
      const nameInput = screen.getByPlaceholderText('Last name');
      await userEvent.type(nameInput, text.name.valid, { allAtOnce: true });

      expect(nameInput).toHaveValue(text.name.valid);
      const nameErrorField = nameInput.previousSibling;
      expect(nameErrorField).toBeEmpty();
    });

    test('last name input should work correct with invalid text ', async () => {
      renderWithAuthRedux(<SignUpForm />);
      const nameInput = screen.getByPlaceholderText('Last name');

      await userEvent.type(nameInput, text.name.invalid, { allAtOnce: true });
      expect(nameInput).toHaveValue(text.name.invalid);
      const nameErrorField = nameInput.previousSibling;
      expect(nameErrorField).toHaveTextContent(text.name.errorMsg);

      userEvent.clear(nameInput);
      expect(nameInput).toHaveValue('');
      expect(nameErrorField).toHaveTextContent(text.name.emptyMsg);
    });

    test('last name input should work correct when invalid text change for valid', async () => {
      renderWithAuthRedux(<SignUpForm />);
      const nameInput = screen.getByPlaceholderText('Last name');

      await userEvent.type(nameInput, text.name.invalid, { allAtOnce: true });
      expect(nameInput).toHaveValue(text.name.invalid);
      const nameErrorField = nameInput.previousSibling;
      expect(nameErrorField).toHaveTextContent(text.name.errorMsg);

      userEvent.clear(nameInput);
      await userEvent.type(nameInput, text.name.valid, { allAtOnce: true });
      expect(nameInput).toHaveValue(text.name.valid);
      expect(nameErrorField).toBeEmpty();
    });

    test('submit button should be enabled  if all fields have valid text', async () => {
      renderWithAuthRedux(<SignUpForm />);
      const passwordInput = screen.getByPlaceholderText('Password');
      const confirmInput = screen.getByPlaceholderText('Confirm password');
      const emailInput = screen.getByPlaceholderText('Email');
      const lastNameInput = screen.getByPlaceholderText('Last name');
      const firstNameInput = screen.getByPlaceholderText('First name');
      const button = screen.getByText('Continue');

      expect(button).toBeDisabled();
      await userEvent.type(emailInput, text.email.valid, { allAtOnce: true });
      await userEvent.type(passwordInput, text.password.valid, { allAtOnce: true });
      await userEvent.type(confirmInput, text.confirm.valid, { allAtOnce: true });
      await userEvent.type(lastNameInput, text.name.valid, { allAtOnce: true });
      await userEvent.type(firstNameInput, text.name.valid, { allAtOnce: true });
      expect(button).toBeEnabled();
    });

    test('submit button should be disabled if at least one field have invalid text', async () => {
      renderWithAuthRedux(<SignUpForm />);
      const passwordInput = screen.getByPlaceholderText('Password');
      const confirmInput = screen.getByPlaceholderText('Confirm password');
      const emailInput = screen.getByPlaceholderText('Email');
      const lastNameInput = screen.getByPlaceholderText('Last name');
      const firstNameInput = screen.getByPlaceholderText('First name');
      const button = screen.getByText('Continue');

      expect(button).toBeDisabled();
      await userEvent.type(emailInput, text.email.valid, { allAtOnce: true });
      await userEvent.type(passwordInput, text.password.valid, { allAtOnce: true });
      await userEvent.type(confirmInput, text.confirm.valid, { allAtOnce: true });
      await userEvent.type(lastNameInput, text.name.invalid, { allAtOnce: true });
      await userEvent.type(firstNameInput, text.name.valid, { allAtOnce: true });
      expect(button).toBeDisabled();

      userEvent.clear(lastNameInput);
      userEvent.clear(emailInput);
      await userEvent.type(emailInput, text.email.invalid, { allAtOnce: true });
      await userEvent.type(lastNameInput, text.name.valid, { allAtOnce: true });
      expect(button).toBeDisabled();
    });

    test('click on submit button should dispatch correct action', async () => {
      const { store } = renderWithAuthRedux(<SignUpForm />);
      const passwordInput = screen.getByPlaceholderText('Password');
      const confirmInput = screen.getByPlaceholderText('Confirm password');
      const emailInput = screen.getByPlaceholderText('Email');
      const lastNameInput = screen.getByPlaceholderText('Last name');
      const firstNameInput = screen.getByPlaceholderText('First name');
      const button = screen.getByText('Continue');

      await userEvent.type(emailInput, text.email.valid, { allAtOnce: true });
      await userEvent.type(passwordInput, text.password.valid, { allAtOnce: true });
      await userEvent.type(confirmInput, text.confirm.valid, { allAtOnce: true });
      await userEvent.type(lastNameInput, text.name.valid, { allAtOnce: true });
      await userEvent.type(firstNameInput, text.name.valid, { allAtOnce: true });
      userEvent.click(button);

      const expectedAction = signUp({
        [EMAIL]: text.email.valid,
        [PASSWORD]: text.password.valid,
        [LAST_NAME]: text.name.valid,
        [FIRST_NAME]: text.name.valid,
      });

      const actions = store.getActions();
      expect(actions).toEqual([expectedAction]);
    });

    test('user can navigate the form from the keyboard', async () => {
      renderWithAuthRedux(<SignUpForm />);
      const passwordInput = screen.getByPlaceholderText('Password');
      const confirmInput = screen.getByPlaceholderText('Confirm password');
      const emailInput = screen.getByPlaceholderText('Email');
      const lastNameInput = screen.getByPlaceholderText('Last name');
      const firstNameInput = screen.getByPlaceholderText('First name');
      const button = screen.getByText('Continue');
      const link = screen.getByText('Sign in');

      expect(document.body).toHaveFocus();

      userEvent.tab();
      expect(emailInput).toHaveFocus();
      await userEvent.type(emailInput, text.email.valid);

      userEvent.tab();
      expect(passwordInput).toHaveFocus();
      await userEvent.type(passwordInput, text.password.valid);

      userEvent.tab();
      expect(confirmInput).toHaveFocus();
      await userEvent.type(confirmInput, text.confirm.valid);

      userEvent.tab();
      expect(firstNameInput).toHaveFocus();
      await userEvent.type(firstNameInput, text.name.valid);

      userEvent.tab();
      expect(lastNameInput).toHaveFocus();
      await userEvent.type(lastNameInput, text.name.valid);

      userEvent.tab();
      expect(button).toHaveFocus();

      userEvent.tab();
      expect(link).toHaveFocus();

      userEvent.tab();
      expect(emailInput).toHaveFocus();
    });
  });
});
