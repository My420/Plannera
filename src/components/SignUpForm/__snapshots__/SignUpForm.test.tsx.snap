// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test SignUpForm component component view test should match snapshot 1`] = `
<DocumentFragment>
  <form
    action="signup.php"
    class="form"
    name="signUpForm"
  >
    <p
      class="title"
    >
      Create account
    </p>
    <div
      class="status"
    >
      <div
        class="container"
        data-testid="container"
      />
    </div>
    <fieldset
      class="signup"
    >
      <label
        class="label input"
        for="email"
      >
        <span
          class="visually-hidden"
        >
          Email
        </span>
        <span
          class="error"
          data-testid="error"
        />
        <input
          class="input "
          id="email"
          name="EMAIL"
          placeholder="Email"
          type="email"
          value=""
        />
      </label>
      <label
        class="label input"
        for="password"
      >
        <span
          class="visually-hidden"
        >
          Password
        </span>
        <span
          class="error"
          data-testid="error"
        />
        <input
          class="input "
          id="password"
          name="PASSWORD"
          placeholder="Password"
          type="password"
          value=""
        />
      </label>
      <label
        class="label input"
        for="confirm"
      >
        <span
          class="visually-hidden"
        >
          Confirm password
        </span>
        <span
          class="error"
          data-testid="error"
        />
        <input
          class="input "
          id="confirm"
          name="CONFIRM"
          placeholder="Confirm password"
          type="password"
          value=""
        />
      </label>
      <label
        class="label input"
        for="firstName"
      >
        <span
          class="visually-hidden"
        >
          First name
        </span>
        <span
          class="error"
          data-testid="error"
        />
        <input
          class="input "
          id="firstName"
          name="FIRST_NAME"
          placeholder="First name"
          type="text"
          value=""
        />
      </label>
      <label
        class="label input"
        for="lastName"
      >
        <span
          class="visually-hidden"
        >
          Last name
        </span>
        <span
          class="error"
          data-testid="error"
        />
        <input
          class="input "
          id="lastName"
          name="LAST_NAME"
          placeholder="Last name"
          type="text"
          value=""
        />
      </label>
    </fieldset>
    <fieldset
      class="controls"
    >
      <button
        class="button "
        disabled=""
        name="signUpSubmit"
        type="submit"
      >
        Continue
      </button>
      <a
        class="link"
        href="/signin"
      >
        Sign in
      </a>
    </fieldset>
  </form>
</DocumentFragment>
`;
